numba/array_analysis.py

```diff
@@ -41,846 +52,1266 @@
+                inds.append(self.obj_to_ind[name])
+            else:
+                inds.append(-1)
+        return tuple(inds)
+
+    def intersect(self, equiv_set):
+        """Overload the intersect method to handle ind_to_var.
+        """
+        newset = super(ShapeEquivSet, self).intersect(equiv_set)
+        ind_to_var = {}
+        for i, objs in newset.ind_to_obj.items():
+            assert(len(objs) > 0)
+            obj = objs[0]
+            assert(obj in self.obj_to_ind)
+            assert(obj in equiv_set.obj_to_ind)
+            j = self.obj_to_ind[obj]
+            k = equiv_set.obj_to_ind[obj]
+            assert(j in self.ind_to_var)
+            assert(k in equiv_set.ind_to_var)
+            varlist = []
+            names = [x.name for x in equiv_set.ind_to_var[k]]
+            for x in self.ind_to_var[j]:
+                if x.name in names:
+                    varlist.append(x)
+            assert(len(varlist) > 0)
+            ind_to_var[i] = varlist
+        newset.ind_to_var = ind_to_var
+        return newset
+
+    def define(self, var):
+        assert(isinstance(var, ir.Var))
+        name = var.name
+        if name in self.defs:
+            self.defs[name] += 1
+            # NOTE: variable being redefined, must invalidate previous equivalences.
+            # Believe it is a rare case, and only happens to scalar accumuators.
+            if name in self.obj_to_ind:
+                i = self.obj_to_ind[name]
+                del self.obj_to_ind[name]
+                self.ind_to_obj[i].remove(name)
+                if len(self.ind_to_obj[i]) == 0:
+                    del self.ind_to_obj[i]
+                assert(i in self.ind_to_var)
+                names = [x.name for x in self.ind_to_var[i]]
+                if name in names:
+                    j = names.index(name)
+                    del self.ind_to_var[i][j]
+                    if len(self.ind_to_var[i]) == 0:
+                        del self.ind_to_var[i]
+                        # no more size variables, need to remove equivalence too
+                        if i in self.ind_to_obj:
+                            for obj in self.ind_to_obj[i]:
+                                del self.obj_to_ind[obj]
+                            del self.ind_to_obj[i]
+        else:
+            self.defs[name] = 1
+
+    def union_defs(self, defs):
+        for k, v in defs.items():
+            if k in self.defs:
+                self.defs[k] = max(self.defs[k], v)
+            else:
+                self.defs[k] = v
 
 class ArrayAnalysis(object):
-    """Analyzes Numpy array computations for properties such as shapes
-    and equivalence classes.
+    """Analyzes Numpy array computations for properties such as
+    shape/size equivalence, and keep track of them on a per-block
+    basis. The analysis should only be run once because it modifies
+    the incoming IR by inserting assertion statements that safeguard
+    parfor optimizations.
```

Comment:
I think the analysis is now idempotent, meaning running it more than once will not alter the IR. I'll add a test case for this.